### PR TITLE
Add working Node.js Native Messaging host

### LINKED ISSUE #27129: Node.js Native Messaging host exits
### MDN URL

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging

### What specific section or headline is this issue about?

Node.js host example here https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging#app_side.

### What information was incorrect, unhelpful, or incomplete?

Node.js host example.

### What did you expect to see?

Working example. 

The current example is incomplete.

### Do you have any supporting links, references, or citations?

Using `process.stdin.on('readable', () => {})` with `process.stdin.read()` is not ideal. Testing this code which uses a similar pattern once `new Array(13107)` (`JSON.stringify(new Array(13107)).length // 65536`) is attempted to be sent from client to host the code will break, the host won't echo the message back, see https://github.com/simov/native-messaging/issues/10. We don't get that far with the code in the documentation because the example is incomplete.

### Do you have anything more you want to share?

This https://github.com/guest271314/native-messaging-nodejs/ is a working example of a Node.js host which sends back 1 MB of data (per protocol).

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/mozilla/add-ons/webextensions/native_messaging`
* MDN URL: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md
* Last commit: https://github.com/mdn/content/commit/d86e14de2e66319fc0a0ec0539a05400dea5a453
* Document last modified: 2023-03-07T05:20:12.000Z

</details>

### PR DESCRIPTION
**Pull request by guest271314**
Fixes https://github.com/mdn/content/issues/27129

### CODE REVIEW

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md [hunk 1]
```diff
@@ -235,67 +235,75 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
...
-    // If the data we have read so far is >= to the size advertised in the header,
-    // it means we have all of the data sent.
-    // We add 4 here because that's the size of the bytes that hold the payloadSize
-    if (stringData.length >= payloadSize + 4) {
-      // Remove the header
-      const contentWithoutSize = stringData.slice(4, payloadSize + 4);
-
-      // Reset the read size and the queued chunks
-      flushChunksQueue();
+#!/usr/bin/env -S /full/path/to/node --max-old-space-size=6 --jitless --expose-gc --v8-pool-size=1
```
Comment (Josh-Cena) [inline files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:241]: This makes the code not directly runnable. What's the consideration for not using `#!/usr/local/bin/node`? Note that if you put it in your package.json `scripts`, npm (and other package managers) is able to map this to the real Node path, even if Node does not live there.

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md [hunk 2]
```diff
@@ -235,67 +235,75 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
...
+    (_, index) => (json.length >> (index * 8)) & 0xff,
+  );
+  let output = new Uint8Array(header.length + json.length);
+  output.set(header, 0);
+  output.set(json, 4);
+  process.stdout.write(output);
+  // Mitigate RSS increasing expotentially for multiple messages
+  // between client and host during same connectNative() connection
+  message = header = output = null;
+  global.gc();
```
Comment (Josh-Cena) [inline files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:292]: I don't understand this part. Why is this necessary and why isn't the default GC reliable?

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md [hunk 3]
```diff
@@ -235,67 +235,75 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
...
+// https://github.com/nodejs/node/issues/43654
+process.env.UV_THREADPOOL_SIZE = 1;
+// Process greater than 65535 length input
+// https://github.com/nodejs/node/issues/6456
+// https://github.com/nodejs/node/issues/11568#issuecomment-282765300
+// https://github.com/nodejs/node/blob/main/doc/api/process.md#a-note-on-process-io
+if (process.stdout._handle) {
+  process.stdout._handle.setBlocking(true);
+}
+// https://github.com/denoland/deno/discussions/17236#discussioncomment-4566134
```
Comment (Josh-Cena) [inline files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:256]: We need to explain what this function solves

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md (hunk 1 above)
Comment (guest271314) [reply files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:241]: In general


- For people who fetch and use only the `node` executable. Contrary to some claims in the wild `node` executable works without any `npm`, `npx`, `node_modules` folder, or package.json file on the sytem;
- For the ability to pass options to the `node` executable on the shebang line

Specifically so users will have to actually write where the `node` executable is located, and don't become conplacent and expect `npm`, which is thrd-party software owned and controlled by GitHub, not Node.js organization, to just magically configure everything for them. Native Messaging _should be_ with hands on development.

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md (hunk 2 above)
Comment (guest271314) [reply files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:292]: `node` will contnually increase RSS on Linux when the options passed to `node` and overwriting variables to `null` and calling `gc()` does not happen. I'm pretty sure I left links in the code to issues where I try to solve that.

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md (hunk 3 above)
Comment (guest271314) [reply files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:256]: `node` is peculiar.  In some cases when `stdout` is greater than 65536 `node` won't write to `stdout`.

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md (hunk 1 above)
Comment (guest271314) [reply files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:241]: > This makes the code not directly runnable. What's the consideration for not using `#!/usr/local/bin/node`?

Depending on what I am doing, I might just get rid of everything except the `node` executable https://github.com/guest271314/download-node-nightly-executable.

Let's say I do something like this to get the nightly Node.js download. 
```
$ cd $HOME
$ mkdir bin
$ cd bin
$ deno
> let osArch = 'linux-x64';
let [node_nightly_build] = await (
  await fetch('https://nodejs.org/download/nightly/index.json')
).json();
let { version, files } = node_nightly_build;
let node_nightly_url = `https://nodejs.org/download/nightly/${version}/node-${version}-${osArch}.tar.gz`;
console.log(node_nightly_url);
# ...
$ wget --show-progress --progress=bar https://nodejs.org/download/nightly/v21.0.0-nightly2023092455fde47b1d/node-v21.0.0-nightly2023092455fde47b1d-linux-x64.tar.gz
# ...
cd $HOME
$ mousepad .profile
# ...
# node-v21.0.0-nightly2023092455fde47b1d-linux-x64
if [ -d "$HOME/bin/node-v21.0.0-nightly2023092455fde47b1d-linux-x64/bin" ] ; then
    PATH="$HOME/bin/node-v21.0.0-nightly2023092455fde47b1d-linux-x64/bin:$PATH"
fi
# ...
$ . ~/.profile
```

I never really _install_ Node.js onto the system. Don't have to.

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md (hunk 3 above)
Comment (guest271314) [reply files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:256]: @Josh-Cena For completeness, another implementation using `node` executable https://github.com/simov/native-messaging/blob/8e99d2a345ae94426a502d05aa5d57b966f6bc78/protocol.js.

Comment (guest271314) [reply files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:256]: @Josh-Cena What's the hold up for merging now?

**Review by Josh-Cena — APPROVED** [review]
Sorry, I thought someone else would come around. This is a much more readable and modern example than the original one, so I'm going to merge it. Thank you for all the work!

**Review by guest271314 — COMMENTED** [review]
Any reason you removed the main function?

**Comment by Josh-Cena** [comment]
@guest271314 This is already an ESM file. We can use top-level await.

**Comment by guest271314** [comment]
Yes, that is true. However, we might want to run other code inside of `main()` before we begin the `while` loop, and/or run other code after we call `main()`. Nothing is going to happen after that `while` loop without including the code in an asynchronous function. Just some functionality and extensibility to consider. We don't gain anything technically by removing the `main` function.

**Comment by Josh-Cena** [comment]
I'm not sure what you mean—if you want to execute anything before the while, just put the code before the while. If you want to do something crazy complicated, just add the wrapper back. For the purpose of this example, without the async function is better.

**Comment by guest271314** [comment]
> For the purpose of this example, without the async function is better.

You havn't explained why you think without the `async` function is "better", which is a subjective term anyway, other than the fact top-level `await` is supported in Ecmascript Modules.

Was just curious why you chose to remove the function. Not something to create controversy over, anyway. Have a great day.

File: files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md [hunk 4]
```diff
@@ -235,67 +235,51 @@ Each message is serialized using JSON, UTF-8 encoded and is preceded with an uns
...
-    // Read all of the available data
-    while ((chunk = process.stdin.read()) !== null) {
-      chunks.push(chunk);
-    }
+async function sendMessage(message) {
+  const header = new Uint32Array([message.length]);
+  const stdout = await fs.open(`/proc/${process.pid}/fd/1`, "w");
+  await stdout.write(header);
+  await stdout.write(message);
+  await stdout.close();
```
Comment (epragalakis) [inline files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:271]: hey @Josh-Cena 
long shot, but do you remember on why you had to do the above (lines 266-271)
instead of doing a `console.log` or `process.stdout.write` ?

**Comment by Josh-Cena** [comment]
This was already there in @guest271314's version, and I don't know why. I would assume `process.stdout` is more portable than assuming the place of the FD, so if you want to change that I'm open to reviewing another PR.
